File: android/src/main/java/com/reactnativestripesdk/checkout/CheckoutControllerRegistry.kt
package com.reactnativestripesdk.checkout

import androidx.annotation.MainThread
import com.facebook.react.bridge.UiThreadUtil
import java.util.UUID

/**
 * Resources owned by one Checkout controller registered with the bridge.
 *
 * Concrete instances own the native Checkout object, its stable Payment Element and presenter,
 * observation jobs, and pending operations.
 */
internal interface CheckoutControllerInstance {
  /** Releases all resources owned by this instance. Must be idempotent and run on the UI thread. */
  @MainThread
  fun destroy()
}

/** Stores Checkout controllers behind opaque identifiers used by JavaScript. */
internal class CheckoutControllerRegistry(
  private val controllerIdFactory: () -> String = { UUID.randomUUID().toString() },
) {
  private data class Entry(
    val instance: CheckoutControllerInstance,
    var eventSequence: Int = 0,
  )

  private val entries = mutableMapOf<String, Entry>()

  val size: Int
    @MainThread
    get() {
      UiThreadUtil.assertOnUiThread()
      return entries.size
    }

  @MainThread
  fun register(instance: CheckoutControllerInstance): String {
    UiThreadUtil.assertOnUiThread()

    var controllerId = controllerIdFactory()
    while (entries.containsKey(controllerId)) {
      controllerId = controllerIdFactory()
    }

    entries[controllerId] = Entry(instance)
    return controllerId
  }

  @MainThread
  fun instance(controllerId: String): CheckoutControllerInstance? {
    UiThreadUtil.assertOnUiThread()
    return entries[controllerId]?.instance
  }

  /** Returns null after the controller has been removed. */
  @MainThread
  fun nextEventSequence(controllerId: String): Int? {
    UiThreadUtil.assertOnUiThread()
    val entry = entries[controllerId] ?: return null
    entry.eventSequence += 1
    return entry.eventSequence
  }

  @MainThread
  fun remove(controllerId: String): Boolean {
    UiThreadUtil.assertOnUiThread()
    val instance = entries.remove(controllerId)?.instance
    instance?.destroy()
    return instance != null
  }

  @MainThread
  fun clear() {
    UiThreadUtil.assertOnUiThread()
    val instances = entries.values.map { it.instance }
    entries.clear()

    instances.forEach { it.destroy() }
  }
}

File: ios/Checkout/CheckoutControllerRegistry.swift
import Foundation

/// Resources owned by one Checkout controller registered with the bridge.
///
/// Concrete instances own the native Checkout object, its stable Payment
/// Element and presenter, observation tokens, and pending operations.
protocol CheckoutControllerInstance: AnyObject {
    /// Releases all resources owned by this instance. Implementations must be idempotent.
    func destroy()
}

/// Stores Checkout controllers behind opaque identifiers used by JavaScript.
final class CheckoutControllerRegistry {
    private struct Entry {
        let instance: any CheckoutControllerInstance
        var eventSequence = 0
    }

    private let lock = NSLock()
    private let makeControllerId: () -> String
    private var entries: [String: Entry] = [:]

    init(makeControllerId: @escaping () -> String = { UUID().uuidString }) {
        self.makeControllerId = makeControllerId
    }

    var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return entries.count
    }

    @discardableResult
    func register(_ instance: any CheckoutControllerInstance) -> String {
        lock.lock()
        defer { lock.unlock() }

        var controllerId = makeControllerId()
        while entries[controllerId] != nil {
            controllerId = makeControllerId()
        }

        entries[controllerId] = Entry(instance: instance)
        return controllerId
    }

    func instance<Instance: CheckoutControllerInstance>(
        for controllerId: String,
        as _: Instance.Type = Instance.self
    ) -> Instance? {
        lock.lock()
        defer { lock.unlock() }
        return entries[controllerId]?.instance as? Instance
    }

    /// Returns the next sequence number for an event from this controller.
    /// Returns `nil` after the controller has been removed.
    func nextEventSequence(for controllerId: String) -> Int? {
        lock.lock()
        defer { lock.unlock() }

        guard var entry = entries[controllerId] else {
            return nil
        }

        entry.eventSequence += 1
        entries[controllerId] = entry
        return entry.eventSequence
    }

    @discardableResult
    func remove(controllerId: String) -> Bool {
        let instance: (any CheckoutControllerInstance)?

        lock.lock()
        instance = entries.removeValue(forKey: controllerId)?.instance
        lock.unlock()

        instance?.destroy()
        return instance != nil
    }

    func removeAll() {
        let instances: [any CheckoutControllerInstance]

        lock.lock()
        instances = entries.values.map(\.instance)
        entries.removeAll()
        lock.unlock()

        instances.forEach { $0.destroy() }
    }
}
